pkg/controllermanager/controller/shootstate/add.go

```diff
@@ -0,0 +1,115 @@
+		).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: ptr.Deref(r.Config.ConcurrentSyncs, 0),
+		}).
+		Watches(
+			&gardencorev1beta1.Shoot{},
+			handler.EnqueueRequestsFromMapFunc(r.MapShootToShootState),
+			builder.WithPredicates(r.ShootPredicates()),
+		).Complete(r)
+}
+
+// MapShootToShootState maps a Shoot object to ShootState reconciliation request.
+func (r *Reconciler) MapShootToShootState(_ context.Context, obj client.Object) []reconcile.Request {
+	shoot, ok := obj.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return nil
+	}
+
+	namespacedName := types.NamespacedName{
+		Name:      shoot.Name,
+		Namespace: shoot.Namespace,
+	}
+	return []reconcile.Request{{NamespacedName: namespacedName}}
+}
+
+// ShootPredicates returns predicates for Shoot requests acceptance.
+func (r *Reconciler) ShootPredicates() predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			shoot, ok := e.ObjectNew.(*gardencorev1beta1.Shoot)
+			if !ok || shoot.Status.LastOperation == nil {
+				return false
+			}
+
+			shootOld, ok := e.ObjectOld.(*gardencorev1beta1.Shoot)
+			if !ok || shootOld.Status.LastOperation == nil {
+				return false
+			}
+
+			oldLastOpType := shootOld.Status.LastOperation.Type
+			newLastOpType := shoot.Status.LastOperation.Type
+			newLastOpState := shoot.Status.LastOperation.State
+
+			isOldLastOpReconcile := oldLastOpType == gardencorev1beta1.LastOperationTypeReconcile
+			isOldLastOpMigrate := oldLastOpType == gardencorev1beta1.LastOperationTypeMigrate
+			isOldLastOpRestore := oldLastOpType == gardencorev1beta1.LastOperationTypeRestore
+
+			isNewLastOpReconcile := newLastOpType == gardencorev1beta1.LastOperationTypeReconcile
+			isNewLastOpMigrate := newLastOpType == gardencorev1beta1.LastOperationTypeMigrate
+			isNewLastOpRestore := newLastOpType == gardencorev1beta1.LastOperationTypeRestore
+			isNewLastOpSucceeded := newLastOpState == gardencorev1beta1.LastOperationStateSucceeded
+
+			isMigrating := false
+			if isOldLastOpReconcile && isNewLastOpMigrate {
+				// Shoot last operation gets updated from Reconcile to Migrate type.
+				isMigrating = true
+			} else if isOldLastOpMigrate && isNewLastOpRestore {
+				// Shoot last operation gets updated from Migrate to Restore type.
+				isMigrating = true
+			} else if isOldLastOpRestore && isNewLastOpRestore && isNewLastOpSucceeded {
+				// Shoot last operation gets updated from Restore to Restore type with Succeeded state.
+				isMigrating = true
+			} else if isOldLastOpRestore && isNewLastOpReconcile {
+				// Shoot last operation gets updated from Restore to Reconcile.
+				isMigrating = true
+			}
+
+			isShootStatePresent := true
+			shootState := &gardencorev1beta1.ShootState{}
+			ctx := context.Background()
+			if err := r.Client.Get(ctx, client.ObjectKeyFromObject(shoot), shootState); err != nil {
+				isShootStatePresent = false
```

Comment:
This sets `isShootStatePresent=false` for any kind of read error. Is this really want we want?

Reply:
Thanks for taking the time to review the logic and bring up the discussion! 

Since the reconciliation logic performs an identical check already, I was thinking to remove it from the predicate and leave it considering only the `.status.lastOperation.Type` transitions. This way we'll have all non-`NotFound` error be handled by the controller itself.

Please let me know your take on it.

Reply:
The way it is implemented now, `Get` is executed for every shoot update which is not something that we want, I think. 
I'm in favour of only considering the `.status.lastOperation.Type` transitions in the predicate.

You could move the `Get` to the mapper function where the context is already passed. Any errors there should be logged, though. For inspiration, check https://github.com/plkokanov/gardener/blob/ea98f76def8b71b7f76b6eb22b3511c7fbada49b/pkg/controllermanager/controller/shoot/conditions/add.go#L79-L108.